File: Android/app/src/main/java/com/derekgillett/sudoku/generator/DailyPuzzle.kt
package com.derekgillett.sudoku.generator

import com.derekgillett.sudoku.model.Difficulty
import java.time.LocalDate

/**
 * Daily puzzle naming and seeding. Same date → same ID and same seed →
 * same puzzle on every device. Currently fixed to a single difficulty.
 *
 * Local-calendar based — what counts as "today" is the user's device date.
 */
object DailyPuzzle {
    val difficulty: Difficulty = Difficulty.MEDIUM

    /** Stable ID for a calendar date: YYYYMMDD packed into an Int. */
    fun id(date: LocalDate = LocalDate.now()): Int {
        return date.year * 10000 + date.monthValue * 100 + date.dayOfMonth
    }

    /**
     * Seed for the deterministic RNG. Spread the date bits through a large
     * multiplier so consecutive days produce visibly different generator
     * outputs.
     */
    fun seed(date: LocalDate = LocalDate.now()): Long {
        // Multiplier is the same bit pattern as the iOS version. Kotlin's
        // Long is signed and 0x9E3779B97F4A7C15 exceeds Long.MAX_VALUE, so
        // we write it as a ULong literal and reinterpret bits to Long. The
        // multiplication itself wraps mod 2^64, matching iOS unchecked
        // overflow operators.
        return id(date).toLong() * 0x9E3779B97F4A7C15uL.toLong()
    }
}

File: iOS/Sudoku/DailyPuzzle.swift
//
//  DailyPuzzle.swift
//  Sudoku
//

import Foundation

/// Daily puzzle naming and seeding. Same date → same ID and same seed →
/// same puzzle on every device. Currently fixed to a single difficulty; if
/// we want per-tier dailies later, the seed can fold the difficulty in.
enum DailyPuzzle {
    static let difficulty: Difficulty = .medium

    /// Stable ID for a calendar date: YYYYMMDD packed into an Int. Local
    /// calendar — what counts as "today" is the user's device date.
    static func id(for date: Date) -> Int {
        let cal = Calendar.current
        let c = cal.dateComponents([.year, .month, .day], from: date)
        let y = c.year ?? 2026
        let m = c.month ?? 1
        let d = c.day ?? 1
        return y * 10000 + m * 100 + d
    }

    /// Seed for the deterministic RNG. Spread the date bits through a
    /// large multiplier so consecutive days produce visibly different
    /// generator outputs.
    static func seed(for date: Date) -> UInt64 {
        UInt64(id(for: date)) &* 0x9E37_79B9_7F4A_7C15
    }
}
